ppcs/ppcTODO-attributes-v2.md

```diff
@@ -0,0 +1,236 @@
+
+* Only lexical or package variables, and subroutines, support attribute syntax. It is not possible to declare attributes on other entities such as packages, or subroutine parameters.
+
+* Attributes on anonymous subroutines are invoked only once at compiletime on the protosub, before any "clonecv" operation that would make a real closure. It cannot perform any behaviour after this time, and does not get to see the real closure CVs that Perl code gets references to.
+
+* Attribute handling is looked up via the package hierarchy expressed in `@ISA`, rather than by lexical scope.
+
+* Third-party attributes are implemented by providing a single `MODIFY_*_ATTRIBUTES` shouty-named method in a package, which is required to understand all the attributes at once. There is no standard mechanism to create individual attributes independently.
+
+Many of these restrictions come from the way that third-party attributes are provided in the perl core, quite apart from its own built-in handling. Built-in attributes get to run their logic much earlier in the parser.
+
+It is the aim of this specification to provide a better and more flexible way for third-party modules to declare and use attributes to allow authors to declare more interesting behaviours on elements of their code. These extensions come in the form of additional elements of the interpreter that are visible to XS modules.
+
+## Rationale
+
+(explain why the (following) proposed solution will solve it)
+
+## Specification
+
+Attributes defined by this specification will be lexical in scope, much like that of a `my` variable, `my sub` function, or any of the builtin function exports provided by the `builtin` module. This provides a clean separation of naming within the code.
+
+An attribute is defined primarily by a C callback function, to be invoked by the parser _as soon as_ it has finished parsing the declaration syntax. This callback function will be passed the target (i.e. the item to which the attribute is being attached), and the optional contents of the parentheses used as an argument to the attribute. There is no interesting return value from this callback function.
+
+An optional second callback function can be provided, for the purpose of parsing the incoming text from the source code into a value that the main apply function will use. Experience with the meta-programming layer in `Object::Pad` suggests this is useful, as often when meta-programming it is inconvenient to have to represent the parameters to an attribute as a flat text string.
+
+Pointers to these two functions are found by a structure that is associated with the name of the new attribute, perhaps defined as the following (though exact type is still an open issue; see below):
+
+```c
+struct PerlAttributeDefinition
+{
+    U32 ver;
+    U32 flags;
+    SV * (*parse)(pTHX_ const SV *text, void *data);   /* optional, may be NULL */
```

Comment:
what about change `parse` to `custom_grammar` returning something that may or may not be `OP *` ? with following behaviour:

- when perl starts:
  - push grammar stack `unbalanced` grammar
  - push grammar stack `perl` grammar
- when encountering places where switch to known custom grammar is needed
  - push grammar stack `known custom grammar` (custom grammar can be `perl` as well)

there already is similar mechanism in `toke.c` - `Perl_lex_start`. This mechanism will just provide generic API to that plus it will provide `Perl` grammar itself as custom grammar.

That itself allows to implement `use VERSION` as custom grammar which may allow to introduce changes will require incompatible changes to `perly.y`.

Example: hypothetical import syntax may reuse this attribute mechanism for itself (by introducing custom attribute context)
```
use builtin
  :functions (
    foo :as (bar)
    baz
  )
  :attributes (magic)
  ;
```